### PR TITLE
[SPARK-30049][SQL] SQL fails to parse when comment contains an unmatched quote character.

### PR DESCRIPTION
**Pull request by javierivanov**
### What changes were proposed in this pull request?

A SQL statement that contains a comment with an unmatched quote character can lead to a parse error:
- Added a insideComment flag in the splitter method to avoid checking single and double quotes within a comment:
```
spark-sql> SELECT 1 -- someone's comment here
         > ;
Error in query: 
extraneous input ';' expecting <EOF>(line 2, pos 0)

== SQL ==
SELECT 1 -- someone's comment here
;
^^^
```

### Why are the changes needed?

This misbehaviour was not present on previous spark versions.

### Does this PR introduce any user-facing change?

- No

### How was this patch tested?

- New tests were added.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @wangyum FYI

**Comment by tgravescs** [comment]
I haven't had time to test your patch but did you test this with just the spark-shell and spark.sql(""" """) syntax as well? I didn't think that went through the SparkSQLCliDriver and that case also fails.

**Comment by kiszk** [comment]
ping @javierivanov

**Comment by javierivanov** [comment]
Sorry for my delay here...
spark.sql will parse only 1 SQL statement, hence the `;` is not accepted by SQL function.
```
scala> sql("""SELECT 1;""")
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input ';' expecting <EOF>(line 1, pos 8)

== SQL ==
SELECT 1;
--------^^^
```
In this case the `;` is being catch in the comment.
```
scala> sql("""SELECT 1 -- someone's comment here;""")
res1: org.apache.spark.sql.DataFrame = [1: int]

```
The SQLDriver is parsing SQLStatements into a list, and foreach calls the sql context to interpret each. 

This was failing because the parser was still detecting quotes inside of comments and passed the sql statement like this:

```
scala> sql("""SELECT 1 -- someone's comment here
     | ;""")
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input ';' expecting <EOF>(line 2, pos 0)

== SQL ==
SELECT 1 -- someone's comment here
;
^^^
```
Let me know for any comments 👍

**Comment by maropu** [comment]
yea, right. Spark doesnt support multi-line queries in `spark.sql`: (related jira: https://issues.apache.org/jira/browse/SPARK-24260)

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 1]
```diff
@@ -509,24 +509,40 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
...
         }
-      } else if (line.charAt(index) == '\"') {
+      } else if (line.charAt(index) == '\"' && !insideComment) {
         // take a look to see if it is escaped
         if (!escape) {
           // flip the boolean variable
           insideDoubleQuote = !insideDoubleQuote
         }
+      } else if (line.charAt(index) == '-') {
+        val hasNext: Boolean = index + 1 < line.length
```
Comment (maropu) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala]: nit: `val hasNext = index + 1 < line.length`

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 2]
```diff
@@ -400,4 +400,21 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
+
+  test("SPARK-30049 Should not complaint for quotes in commented with multi-lines") {
+    runCliWithin(1.minute)(
+      """SELECT concat('test', 'comment') -- someone's comment here \\
+        |comment continues here with single ' quote
+        | extra '
+        |;""".stripMargin -> "testcomment"
+    )
+  }
+
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: nit: remove this unnecessary blank.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 3]
```diff
@@ -400,4 +400,21 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
+  test("SPARK-30049 Should not complaint for quotes in commented with multi-lines") {
+    runCliWithin(1.minute)(
+      """SELECT concat('test', 'comment') -- someone's comment here \\
+        |comment continues here with single ' quote
+        | extra '
+        |;""".stripMargin -> "testcomment"
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Can you add a test case having spaces in the head, e.g., `|     ;""".stripMargin`?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 4]
```diff
@@ -509,24 +509,40 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
...
+        if (insideDoubleQuote || insideSingleQuote || insideComment) {
+          // ignore
```
Comment (maropu) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala]: Can you add more comments here about which case we can ignore?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 2 above)
Comment (javierivanov) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: @maropu Not sure, which blank?

**Comment by tgravescs** [comment]
> yea, right. Spark doesnt support multi-line queries in spark.sql: (related jira: https://issues.apache.org/jira/browse/SPARK-24260)

ah ok, that makes sense, thanks

**Comment by maropu** [comment]
ping @wangyum

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 5]
```diff
@@ -509,24 +509,43 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
...
+          // Ignores '-' in any case of quotes or comment.
+          // Avoids to start a comment(--) within a quoted segment or already in a comment.
+          // Sample query: select "quoted value --"
+          //                                    ^^ avoids starting a comment if it's inside quotes.
+        } else if (hasNext && line.charAt(index + 1) == '-') {
+          // ignore quotes and ;
+          insideComment = true
+          // ignore eol
+          endIndex = index
+        } else {
```
Comment (tgravescs) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala]: we can just remove this else {

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 6]
```diff
@@ -509,24 +509,43 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
...
+          // continue
+        }
+
```
Comment (tgravescs) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala]: nit: remove the extra new line here

**Comment by tgravescs** [comment]
@wangyum any comments?

**Review by tgravescs — APPROVED** [review]

**Comment by tgravescs** [comment]
merged to master and branch-3.0, thanks @javierivanov

**Comment by gatorsmile** [comment]
cc @juliuszsompolski

**Comment by dongjoon-hyun** [comment]
Hi, All.
It seems that https://issues.apache.org/jira/browse/SPARK-31102 is reported due to this by @wangyum . Could you take a look?

**Pull request closed without merging**